core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2242,70 +2252,110 @@ private[spark] class DAGScheduler(
-              case shuffleServiceLoc =>
-                // Sends async request to shuffle service to finalize shuffle merge on that host.
-                // Since merge statuses will not be registered in this case,
-                // we pass a no-op listener.
-                shuffleClient.finalizeShuffleMerge(shuffleServiceLoc.host,
-                  shuffleServiceLoc.port, shuffleId, shuffleMergeId,
-                  new MergeFinalizerListener {
-                    override def onShuffleMergeSuccess(statuses: MergeStatuses): Unit = {
-                    }
+      val scheduledFutures =
+        if (!registerMergeResults) {
+          results.foreach(_.set(true))
+          // Finalize in separate thread as shuffle merge is a no-op in this case
+          stage.shuffleDep.getMergerLocs.map {
+            case shuffleServiceLoc =>
+              // Sends async request to shuffle service to finalize shuffle merge on that host.
+              // Since merge statuses will not be registered in this case,
+              // we pass a no-op listener.
+              shuffleSendFinalizeRpcExecutor.submit(new Runnable() {
+                override def run(): Unit = {
+                  shuffleClient.finalizeShuffleMerge(shuffleServiceLoc.host,
+                    shuffleServiceLoc.port, shuffleId, shuffleMergeId,
+                    new MergeFinalizerListener {
+                      override def onShuffleMergeSuccess(statuses: MergeStatuses): Unit = {
+                      }
 
-                    override def onShuffleMergeFailure(e: Throwable): Unit = {
-                    }
-                  })
-            }
-          }
-        }, 0, TimeUnit.SECONDS)
-      } else {
-        stage.shuffleDep.getMergerLocs.zipWithIndex.foreach {
-          case (shuffleServiceLoc, index) =>
-            // Sends async request to shuffle service to finalize shuffle merge on that host
-            // TODO: SPARK-35536: Cancel finalizeShuffleMerge if the stage is cancelled
-            // TODO: during shuffleMergeFinalizeWaitSec
-            shuffleClient.finalizeShuffleMerge(shuffleServiceLoc.host,
-              shuffleServiceLoc.port, shuffleId, shuffleMergeId,
-              new MergeFinalizerListener {
-                override def onShuffleMergeSuccess(statuses: MergeStatuses): Unit = {
-                  assert(shuffleId == statuses.shuffleId)
-                  eventProcessLoop.post(RegisterMergeStatuses(stage, MergeStatus.
-                    convertMergeStatusesToMergeStatusArr(statuses, shuffleServiceLoc)))
-                  results(index).set(true)
+                      override def onShuffleMergeFailure(e: Throwable): Unit = {
+                        if (e.isInstanceOf[IOException]) {
+                          logInfo(s"Failed to connect external shuffle service " +
+                            s"${shuffleServiceLoc.hostPort}")
+                          blockManagerMaster.removeShufflePushMergerLocation(shuffleServiceLoc.host)
```

Comment:
Why are we doing this? Network issues can be transient. We don't have retries for sending finalize RPC which is fine because push/merge operation is best effort. However, why do we decide remove the host from shuffle push merger location. I don't recommend clubbing these changes in the bug fix.

Reply:
Because we don't want those nodes that failed to connect to be selected as merge locations again. 
The merge locations come from two parts, the active block managers and the shuffle services that current app registered an executor in the past. We will remove node that failed to connect from the second part.
So if the merge ESS is still active, it will still can be selected as merge location as it is in the first part. Or it will not be selected as merge location if it is down.

Reply:
When the connection fails to get established, it will not be always because ESS is down. It could be transient network failure. We don't retry connecting to it here so I don't think based on the connection failure we should remove the merger.

> Because we don't want those nodes that failed to connect to be selected as merge locations again.

This may happen any ways. Since the ESS is down, there will be fetch failures against it and the current code in DAGScheduler will handle removing the unreachable host from merger locations.

Reply:
Resolving, as this was reverted.